### DISCUSSION TITLE
Additional cost based on total distance

### DISCUSSION THREAD

**Discussion by hrushk** [google/or-tools#2388] [Routing (and legacy CP) questions]
**What language and solver does this apply to?**
Python

**Describe the problem you are trying to solve.**
(VRPTW) Currently working in adding a cost based on total distance travelled by the vehicle --> Rounding off the travel time to the next multiple of 10.

**Describe the solution you'd like**
A way to modify the optimization objective as per the need. Specifically modify the travel cost of each vehicle to the nearest multiple of 10.

**Describe alternatives you've considered**
I tried implementing the above using **AddDistanceDependentDimension**, but I could not find a correct method to do so in Python. Is there any example that I can refer to.

**Additional context**
What I am trying to do is to optimize route cost in unit of 10.
Eg. If a vehicle travel from time 11 to 34, the usual cost should be 23 units, but since I want to do in units of 10, the cost will be from 10 to 30 which is 20 units. Can anyone please tell me the correct way to do this?
Any help would be really appreciated.

**Comment by lperron** [comment]
1) this is not an issue
2) the request is unreasonable. Why don't you post-process the cost to round it to the nearest multiple of 10 ?

**Reply by hrushk** [reply]
Pardon me for putting this on wrong section.
I want to round off the total route cost only.  In case I do this in `time_callback `, it will affect the individual travel time which is not what I want.
I am new to VRPTW (following example: https://developers.google.com/optimization/routing/vrptw#python_6) and don't know a correct way to use entire route cost instead of individual time_callback between from and to nodes. Could you perhaps guide me to a proper implementation of entire route cost?

**Reply by lperron** [reply]
the cost will be the entire route cost as the sum of the individual time_callback.
What is your question again?

**Reply by hrushk** [reply]
Sorry for brevity. Allow me explain what I am trying to do.
if a route for a vehicle is 0-->2-->5-->7-->0 [0 being the depot],
```python
# Create and register a transit callback.
    def time_callback(from_index, to_index):
        """Returns the travel time between the two nodes."""
        # Convert from routing variable Index to time matrix NodeIndex.
        from_node = manager.IndexToNode(from_index)
        to_node = manager.IndexToNode(to_index)
        return data['time_matrix'][from_node][to_node]

    transit_callback_index = routing.RegisterTransitCallback(time_callback)
    
    # Define cost of each arc.
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)`
```
Using this the time_callback takes from and to node to calculate the cost between each pair of visits by the vehicle. The cost for entire route will indeed be the sum of individual cost.  But as this time_callback takes to and from nodes as inputs, is there a way to modify this one to only include the first and last node and then feed the modified cost to SetArcCostEvaluatorOfAllVehicles?
 You mentioned that I can sum of individual time_callback, but then how do I feed it to SetArcCostEvaluatorOfAllVehicles as it takes to and from node irrespective of the position of the node in the route.

> def SetArcCostEvaluatorOfAllVehicles(self, evaluator_index: int) -> 'void'
> Sets the cost function of the model such that the cost of a segment of a route between node 'from' and 'to' is evaluator(from, to), whatever the route or vehicle performing the route.

**Comment by lperron** [comment]
You can have as many dimensions as you want.
One will be time, and have time windows if needed on them.

One will be your modified cost structure, and can be used for the objective.

Does it answer your question?
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00

**Reply by Mizux** [reply]
if your route is `0-->2-->5-->7-->0` the total arc cost will be the **sum** of `[T(0, 2), T(2, 5), T(5, 7), T(7, 0) ]` with `T` being your registered transit callback, so cost "from first" and "to last" node (i.e. depot) will be take into account...

**Reply by hrushk** [reply]
Thank you both for your replies. I think I need to search more about how dimensions work and how to manipulate them. I will try to build a solution based on your feedback. Thanks a lot.
